src/compiler/crystal/macros/macros.cr

```diff
@@ -49,6 +53,25 @@ class Crystal::Program
     interpreter.to_s
   end
 
+  # Cache of free_vars of GenericInstanceType for macro expansion.
+  @free_vars_cache = {} of GenericInstanceType => Hash(String, TypeVar)
+
+  private def macro_free_vars_for_type(type)
+    unless free_vars = @free_vars_cache[type]?
```

Comment:
Maybe use the following syntax?
```cr
@free_vars_cache[type] ||= begin
  # build the value for this *type* here
end
```
This will get the value associated with `type`, and if there is none, build the value and associate it with `type` in the hash, and return it.